### PR TITLE
Add autoconfiguration for client-side observation interceptor

### LINKED ISSUE #51: Support for client side observability
The server side works fine now via the autconfigured interceptor but there's nothing for clients yet. Probably best to wait for #52, or use this as a way to investigate design work for that issue.

### PR DESCRIPTION
**Pull request by NotBjoggisAtAll**
[resolves #51]

### CODE REVIEW

File: spring-grpc-spring-boot-autoconfigure/src/main/java/org/springframework/grpc/autoconfigure/client/GrpcClientObservationAutoConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.context.annotation.Bean;
+import org.springframework.grpc.client.GlobalClientInterceptor;
+
+@AutoConfiguration(
+		afterName = "org.springframework.boot.actuate.autoconfigure.observation.ObservationAutoConfiguration")
+@ConditionalOnClass({ ObservationRegistry.class, ObservationGrpcClientInterceptor.class })
+@ConditionalOnBean(ObservationRegistry.class)
+@ConditionalOnProperty(name = "spring.grpc.client.observation.enabled", havingValue = "true", matchIfMissing = true)
```
Comment (onobc) [inline spring-grpc-spring-boot-autoconfigure/src/main/java/org/springframework/grpc/autoconfigure/client/GrpcClientObservationAutoConfiguration.java:31]: Do you mind adding an entry to the [config props additional metadata](https://github.com/spring-projects-experimental/spring-grpc/blob/729940d62c15fb198e5e3232fec7f06031f80ede/spring-grpc-spring-boot-autoconfigure/src/main/resources/META-INF/additional-spring-configuration-metadata.json#L21)?

File: spring-grpc-spring-boot-autoconfigure/src/test/java/org/springframework/grpc/autoconfigure/client/GrpcClientObservationAutoConfigurationTest.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+	@Test
+	void whenObservationPropertyDisabledThenAutoConfigIsSkipped() {
+		this.validContextRunner()
+			.withPropertyValues("spring.grpc.client.observation.enabled=false")
+			.run(context -> assertThat(context).doesNotHaveBean(GrpcClientObservationAutoConfiguration.class));
+	}
+
+	@Test
+	void whenAllConditionsAreMetThenInterceptorConfiguredAsExpected() {
```
Comment (onobc) [inline spring-grpc-spring-boot-autoconfigure/src/test/java/org/springframework/grpc/autoconfigure/client/GrpcClientObservationAutoConfigurationTest.java:80]: I like this test you have added. It checks if the interceptor bean is registered and if its marked as global interceptor. I think the test may be a bit easier to follow by using `hasEntrySatisifying` - something like:
```java
@Test
void whenAllConditionsAreMetThenInterceptorConfiguredAsExpected() {
this.contextRunner().run((context) -> {
	assertThat(context).hasSingleBean(ObservationGrpcClientInterceptor.class);
	assertThat(context.getBeansWithAnnotation(GlobalClientInterceptor.class))
			.hasEntrySatisfying("observationGrpcClientInterceptor", 
					(bean) -> bean.getClass().isAssignableFrom(ObservationGrpcClientInterceptor.class));
});
}
```
Wdyt?

**Review by onobc — CHANGES_REQUESTED** [review]
Thanks for the contribution @NotBjoggisAtAll . nice work. I added a couple minor comments/suggestions, but other than that LGTM.

**Comment by onobc** [comment]
I almost forgot, we have a small section on the server side observability docs [here](https://docs.spring.io/spring-grpc/reference/server.html#_observability).
It would be nice to do the same in the client section.
@NotBjoggisAtAll if you don't get to this that is ok - we can add it during merge etc..

File: spring-grpc-spring-boot-autoconfigure/src/test/java/org/springframework/grpc/autoconfigure/client/GrpcClientObservationAutoConfigurationTest.java (hunk 2 above)
Comment (NotBjoggisAtAll) [reply spring-grpc-spring-boot-autoconfigure/src/test/java/org/springframework/grpc/autoconfigure/client/GrpcClientObservationAutoConfigurationTest.java:80]: Yes I agree. 

I copied the tests from [GrpcServerObservationAutoConfigurationTests](https://github.com/spring-projects-experimental/spring-grpc/blob/729940d62c15fb198e5e3232fec7f06031f80ede/spring-grpc-spring-boot-autoconfigure/src/test/java/org/springframework/grpc/autoconfigure/server/GrpcServerObservationAutoConfigurationTests.java).
 Do you want me to adjust the test for the server too? Should that be done in a separate PR?

File: spring-grpc-spring-boot-autoconfigure/src/main/java/org/springframework/grpc/autoconfigure/client/GrpcClientObservationAutoConfiguration.java (hunk 1 above)
Comment (NotBjoggisAtAll) [reply spring-grpc-spring-boot-autoconfigure/src/main/java/org/springframework/grpc/autoconfigure/client/GrpcClientObservationAutoConfiguration.java:31]: Yes, I added it below the server version of this property

**Comment by NotBjoggisAtAll** [comment]
I added an section for Observability for the client. I excluded the last part that was written on the server docs because I don't think we have a client sample yet?

I also had to adjust the test a bit. I think an assert inside the lamda is correct? Otherwise it seems that the test passes no matter what I put in `isAssignableFrom()`

Please let me know if there is anything else you want me to do

File: spring-grpc-docs/src/main/antora/modules/ROOT/pages/client.adoc [hunk 3]
```diff
@@ -131,3 +131,8 @@ However, by setting the `mergeWithGlobalInterceptors` parameter on the channel f
 You can use this option if you want to add a per-client interceptor between global interceptors.
 
 IMPORTANT: The per-channel interceptors you pass in must either be bean instances marked with `@Order` or regular objects that implement the `Ordered` interface to be properly merged/ordered with the global interceptors.
+
+== Observability
+
+Spring gRPC provides an autoconfigured interceptor that can be used to provide observability to your gRPC services.
+All you need to do is add Spring Boot actuators to your project, and optionally a bridge to your observability platform of choice (just like https://docs.spring.io/spring-boot/reference/actuator/observability.html[any other Spring Boot application]).
```
Comment (dsyer) [inline spring-grpc-docs/src/main/antora/modules/ROOT/pages/client.adoc:138]: Would actuators be necessary or appropriate for a client app (no server)?

**Comment by dsyer** [comment]
You need to build locally too, to make sure the CI build succeeds. It tells you how to fix it.

File: spring-grpc-docs/src/main/antora/modules/ROOT/pages/client.adoc (hunk 3 above)
Comment (NotBjoggisAtAll) [reply spring-grpc-docs/src/main/antora/modules/ROOT/pages/client.adoc:138]: Yes I agree, I can remove the part talking about Spring Boot actuators and leave only the first line.

If you have more feedback on what else should be stated in the docs please let me know. Writing docs is not my strong side 😅

**Comment by NotBjoggisAtAll** [comment]
> You need to build locally too, to make sure the CI build succeeds. It tells you how to fix it.

Yeah I noticed, I will take a look at it soon

File: spring-grpc-spring-boot-autoconfigure/src/test/java/org/springframework/grpc/autoconfigure/client/GrpcClientObservationAutoConfigurationTest.java (hunk 2 above)
Comment (onobc) [reply spring-grpc-spring-boot-autoconfigure/src/test/java/org/springframework/grpc/autoconfigure/client/GrpcClientObservationAutoConfigurationTest.java:79]: Thats ok @NotBjoggisAtAll , I will update that one separately.

**Review by onobc — APPROVED** [review]

**Comment by dsyer** [comment]
Merged in 4149ed5. Thank you.

**Pull request closed without merging**